### PR TITLE
Migrate deployment reconciling

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
Migrate deployment reconciling

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/resources/ensure_test.go [hunk 1]
```diff
@@ -62,3 +70,142 @@ func BenchmarkEnsureRole(b *testing.B) {
...
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var client controllerruntimeclient.Client
+			if test.existingObject != nil {
+				client = controllerruntimefake.NewFakeClient(test.existingObject)
+			} else {
+				client = controllerruntimefake.NewFakeClient()
+			}
+
+			store := &fakeSingleItemInformer{
```
Comment (mrIncompetent) [inline api/pkg/resources/ensure_test.go:176]: It's really hard to use the controller-runtime informer factory during tests... Which is why I created a really dump testing one...
But hey, its simple and we have tests :tada:

File: api/pkg/resources/zz_generated_reconcile.go [hunk 2]
```diff
@@ -79,6 +79,38 @@ func ReconcileStatefulSets(creators []StatefulSetCreator, namespace string, clie
...
+	return func(existing runtime.Object) (runtime.Object, error) {
+		if existing != nil {
+			return create(existing.(*appsv1.Deployment))
+		}
+		return create(&appsv1.Deployment{})
+	}
+}
+
+// ReconcileDeployments will create and update the Deployments coming from the passed DeploymentCreator slice
+func ReconcileDeployments(creators []DeploymentCreator, namespace string, client ctrlruntimeclient.Client, informerFactory ctrlruntimecache.Cache, wrapper ...ObjectModifier) error {
```
Comment (alvaroaleman) [inline api/pkg/resources/zz_generated_reconcile.go]: can we rename `wrapper` to `objectModifiers`? An `objectModifier` is functining as a wrapper but that does't tell anything about its purpose, it took me some time to understand why this is done

File: api/pkg/resources/zz_generated_reconcile.go [hunk 3]
```diff
@@ -79,6 +79,38 @@ func ReconcileStatefulSets(creators []StatefulSetCreator, namespace string, clie
 	return nil
 }
 
+// DeploymentObjectWrapper adds a wrapper so the DeploymentCreator matches ObjectCreator
+// This is needed as golang does not support function interface matching
+func DeploymentObjectWrapper(create DeploymentCreator) ObjectCreator {
```
Comment (alvaroaleman) [inline api/pkg/resources/zz_generated_reconcile.go]: Why is this public?

File: api/pkg/resources/zz_generated_reconcile.go [hunk 4]
```diff
@@ -79,6 +79,38 @@ func ReconcileStatefulSets(creators []StatefulSetCreator, namespace string, clie
...
+		return fmt.Errorf("failed to get Deployment informer: %v", err)
+	}
+
+	for _, create := range creators {
+		createObject := DeploymentObjectWrapper(create)
+		for _, wrap := range wrapper {
+			createObject = wrap(createObject)
+		}
+
+		if err := EnsureObject(namespace, createObject, store, client); err != nil {
```
Comment (alvaroaleman) [inline api/pkg/resources/zz_generated_reconcile.go:106]: Nothing to do with this PR, but in an ideal world, `createObject` would be an interface type that has these funcs:

`
type creatorInterface {
 Creator func(existing runtime.Object) (runtime.Object, error)
 Name string
}
`

Then we wouldn't have to call the creator twice in `EnsureObject`

**Review by alvaroaleman — COMMENTED** [review]
Sorry for the huge delay, but this is difficult to review just because of its sheer size. Very good change thought, I like the -2k LOC :)

File: api/pkg/resources/zz_generated_reconcile.go (hunk 4 above)
Comment (mrIncompetent) [reply api/pkg/resources/zz_generated_reconcile.go:106]: Fun fact: 
I did some work on exactly that for Secrets and ConfigMaps. As soon as this PR is merged, I'll open a new PR :)

File: api/pkg/resources/zz_generated_reconcile.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/resources/zz_generated_reconcile.go]: Renamed with https://github.com/kubermatic/kubermatic/pull/2738/commits/378edee4ec0fd583c074a97d8c6277289cbe5f54

File: api/pkg/resources/zz_generated_reconcile.go (hunk 3 above)
Comment (mrIncompetent) [reply api/pkg/resources/zz_generated_reconcile.go]: Made the func private with https://github.com/kubermatic/kubermatic/pull/2738/commits/378edee4ec0fd583c074a97d8c6277289cbe5f54

**Comment by mrIncompetent** [comment]
@alvaroaleman PTAL

**Comment by mrIncompetent** [comment]
@kron4eg Please have a quick look as I touched some code of the monitoring controller

**Pull request merged**